src/js/core/methods/EthereumSignTypedData.js

```diff
@@ -0,0 +1,172 @@
+            { name: 'metamaskV4Compatibility', type: 'boolean', obligatory: true },
+        ]);
+
+        const path = validatePath(payload.path, 3);
+        const network = getEthereumNetwork(path);
+        this.firmwareRange = getFirmwareRange(this.name, network, this.firmwareRange);
+
+        this.info = getNetworkLabel('Sign #NETWORK typed data', network);
+
+        const { data, metamaskV4Compatibility } = payload;
+
+        this.params = {
+            path,
+            network,
+            data,
+            metamaskV4Compatibility,
+        };
+    }
+
+    async run() {
+        const cmd = this.device.getCommands();
+        const { path: address_n, network, data, metamaskV4Compatibility } = this.params;
+
+        const { types, primaryType, domain, message } = data;
+
+        let response: MessageResponse<
+            | 'EthereumTypedDataStructRequest'
+            | 'EthereumTypedDataValueRequest'
+            | 'EthereumTypedDataSignature',
+        > = await cmd.typedCall(
+            'EthereumSignTypedData',
+            // $FlowIssue typedCall problem with unions in response, TODO: accept unions
+            'EthereumTypedDataStructRequest|EthereumTypedDataValueRequest|EthereumTypedDataSignature',
+            {
+                address_n,
+                primary_type: primaryType,
+                metamask_v4_compat: metamaskV4Compatibility,
+            },
+        );
+
+        // sending all the type data
+        while (response.type === 'EthereumTypedDataStructRequest') {
+            // $FlowIssue disjoint union Refinements not working, TODO: check if new Flow versions fix this
+            const { name: typeDefinitionName } = response.message;
+            const typeDefinition = types[typeDefinitionName];
+            if (typeDefinition === undefined) {
+                throw ERRORS.TypedError(
+                    'Method_InvalidParameter',
+                    `Type ${typeDefinitionName} was not defined in types object`,
+                );
+            }
+            const dataStruckAck: EthereumTypedDataStructAck = {
+                members: typeDefinition.map(({ name, type: typeName }) => ({
+                    name,
+                    type: getFieldType(typeName, types),
+                })),
+            };
+            response = await cmd.typedCall(
+                'EthereumTypedDataStructAck',
+                // $FlowIssue typedCall problem with unions in response, TODO: accept unions
+                'EthereumTypedDataStructRequest|EthereumTypedDataValueRequest|EthereumTypedDataSignature',
+                dataStruckAck,
+            );
+        }
+
+        // sending the whole message to be signed
+        while (response.type === 'EthereumTypedDataValueRequest') {
+            // $FlowIssue disjoint union Refinements not working, TODO: check if new Flow versions fix this
+            const valueRequestMessage: EthereumTypedDataValueRequest = response.message;
```

Comment:
Total nitpick: I would rather destructure response.message directly here on order to avoid defining a new variable `valueRequestMessage`

Reply:
Done in https://github.com/trezor/connect/pull/983/commits/da7872c2e91df03936e2c2c7c9ffa97798237430

My reason for adding it in was because Flow couldn't decode the types properly (it really doesn't like `await cmd.typedCall()` with union types, and flow thinks it might still be a `Promise<>`).

You didn't mention it, but I also removed the `const signatureMessage: EthereumTypedDataSignature = response.message;` in the same commit, since I was also defining a new variable `signatureMessage` instead of destructuring `response.message` directly.